lib/dns.js

```diff
@@ -212,8 +245,22 @@ function lookup(hostname, options, callback) {
     return {};
   }
 
+  const cacheKey = `${hostname}_${family || 'default'}`;
+  const cachedResult = dnsCache[cacheKey];
```

Comment:
A read-through cache approach would be a bit simpler here. essentially something like

```
cache.get(cacheKey, callback, () => {
  // If the cache lookup is a miss, this function is called to do the lookup and cache the result of the lookup.
});
```